Add DATE_TRUNC SQL love letter (#1670)
fishtown-analytics/docs.getdbt.com: website/blog/2022-07-05-date-trunc-sql-love-letter.md

---
title: "DATE_TRUNC SQL function: Why we love it"
description: "The DATE_TRUNC function will truncate a date or time to the first instance for a given date part maintaining a date format. Wordy, wordy, wordy! What does this really mean?"
slug: date-trunc-sql

authors: [kira_furuichi]

tags: [sql magic]
hide_table_of_contents: false

date: 2022-07-05
is_featured: false
---
In general, data people prefer the more granular over the less granular. [Timestamps > dates](https://docs.getdbt.com/blog/when-backend-devs-spark-joy#signs-the-data-is-sparking-joy), daily data > weekly data, etc.; having data at a more granular level always allows you to zoom in. However, you’re likely looking at your data at a somewhat zoomed-out level—weekly, monthly, or even yearly. To do that, you’re going to need a handy dandy function that helps you round out date or time fields.

> The DATE_TRUNC function will truncate a date or time to the first instance for a given date part maintaining a date format. Wordy, wordy, wordy! What does this really mean? If you were to truncate ‘2021-12-13’ out to its month, it would return ‘2021-12-01’.

Reviewer: ```suggestion
The DATE_TRUNC function will truncate a date or time to the first instance of a given date part. Wordy, wordy, wordy! What does this really mean? If you were to truncate `2021-12-13` out to its month, it would return `2021-12-01` (the first day of the month).
```

Below, we’ll outline some of the slight differences in the implementation between some of the data warehouses.

### The DATE_TRUNC function in Snowflake and Databricks

In [Snowflake](https://docs.snowflake.com/en/sql-reference/functions/date_trunc.html) and [Databricks](https://docs.databricks.com/sql/language-manual/functions/date_trunc.html), you can use the DATE_TRUNC function using the following syntax:

```sql
date_trunc(<date_part>, <date/time field>)
```

In these platforms, the `<date_part>` is passed in as the first argument in the DATE_TRUNC function.

### The DATE_TRUNC function in Google BigQuery and Amazon Redshift

In [Google BigQuery](https://cloud.google.com/bigquery/docs/reference/standard-sql/date_functions#date_trunc) and [Amazon Redshift](https://docs.aws.amazon.com/redshift/latest/dg/r_DATE_TRUNC.html), the `<date_part>` is passed in as the first argument and the `<date/time field>` is the second argument.

```sql
date_trunc(<date/time field>, <date part>)
```

> > **Note**

Reviewer: ```suggestion
> **Note:**
```

## A dbt macro to remember

Why Snowflake, Amazon Redshift, Databricks, and Google BigQuery decided to use different implementations of essentially the same function is beyond us and it’s not worth the headache trying to figure that out. Instead of remembering if the`<date_part>` or the `<date/time field>` comes first, (which, let’s be honest, we can literally never remember) you can rely on a dbt Core macro to help you get away from finicky syntax.

With dbt v1.2, [adapters](https://docs.getdbt.com/docs/available-adapters) now support [cross-database macros](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros) to help you write certain functions, like [DATE_TRUNC](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#date_trunc) and [DATEDIFF](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#datediff), without having to memorize sticky function syntax.

> > **Note**

Reviewer: ```suggestion
> **Note:**
```

> Previously, [dbt_utils](https://github.com/dbt-labs/dbt-utils), a package of macros and tests that data folks can use to help write more DRY code in their dbt project, powered cross-database macros. Now, cross-database macros are available **regardless if dbt utils is installed or not.**

Using the [jaffle shop](https://github.com/dbt-labs/jaffle_shop/blob/main/models/orders.sql), a simple dataset and dbt project, you can truncate the `order_date` from the `orders` table using the dbt [DATE_TRUNC macro](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#date_trunc):

```sql
select
	order_id,
	order_date,

> 	{{ date_trunc(‘week’, order_date) }} as order_week,

Reviewer: ```suggestion
	{{ date_trunc("week", "order_date") }} as order_week,
```

> 	{{ date_trunc(‘month’, order_date) }} as order_month,

Reviewer: ```suggestion
	{{ date_trunc("month", "order_date") }} as order_month,
```

> 	{{ date_trunc(‘year’, order_date) }} as order_year

Reviewer: ```suggestion
	{{ date_trunc("year", "order_date") }} as order_year
```

> from {{ ref(‘orders’) }}

Reviewer: ```suggestion
from {{ ref('orders') }}
```
